### ISSUE TITLE
AvailableConditionController doesn't implement proper backoff/retry strategy

### ISSUE THREAD

**Issue by mborsz** [kubernetes/kubernetes#70287]
**What happened**:
When AvailableConditionController is unable to contact a api service it keeps retrying with high number of attempts per seconds (like 80 attempts per seconds) creating significant cpu usage. 
https://gist.github.com/mborsz/c6094ae23f84db10593f1e0a0fd5d38d contains kube-apiserver's logs from one of the attempts.

**What you expected to happen**:
AvailableConditionController should implement a proper backoff/retry strategy in that case.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a cluster
2. Block ssh tunnel access to the nodes with service running
3. Restart kube-apiserver
4. Watch logs and cpu usage of kube-apiserver

**Anything else we need to know?**:

AvailableConditionController seems to be trying to implement some backoff logic, but it either doesn't work or still generates too big load.

**Environment**:
- Kubernetes version (use `kubectl version`): 1.9.7
- Cloud provider or hardware configuration: gke
- OS (e.g. from /etc/os-release):  cos
- Kernel (e.g. `uname -a`): 4.4.111+
- Install tools:
- Others:

/kind bug

**Comment by mborsz** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by wojtek-t** [comment]
@cheftako @mml

**Comment by liggitt** [comment]
cc @deads2k

**Comment by yue9944882** [comment]
@liggitt yeah.. was just surprised.. so the error might be nil in this case?

**Comment by liggitt** [comment]
noticed this was reported against 1.9:

https://github.com/kubernetes/kubernetes/blob/b5ec061c7ab9995a801206ea74f614ced04206d2/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L91

https://github.com/kubernetes/kubernetes/blob/b5ec061c7ab9995a801206ea74f614ced04206d2/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L295-L296

the way the rate limiter was constructed has changed since then, and is now a per-item backoff:

https://github.com/kubernetes/kubernetes/blob/e463d10935c4e4ef683e26c63a5994976addef11/staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go#L91-L97

is this still reproducible in master?

**Comment by mborsz** [comment]
I was able to reproduce this issue using relatively fresh (v1.13.0-alpha.0) version. The retry rate was lower (like 8 retries per second), but the CPU was still high (like ~80%).
When firewall rule used to block ssh tunnel creation was removed, the cpu usage has dropped to nearly 0%.

Logs from v1.13.0-alpha.0 version: https://gist.github.com/mborsz/0590e07ee7d12bdcc369aa3f8637fcb5

**Comment by mml** [comment]
@liggitt your code snippets makes it seem like we should quickly (well, not sure how quickly really) decay to a rate of 2/min, not 8/sec.

Also, do we have way of creating a test that simulates this condition?  I.e. many nodes (they don't have to be real) and a blackhole firewall rule that causes ssh connection attempts to time out.

**Comment by liggitt** [comment]
> @liggitt your code snippets makes it seem like we should quickly (well, not sure how quickly really) decay to a rate of 2/min, not 8/sec.

agree



> Also, do we have way of creating a test that simulates this condition? I.e. many nodes (they don't have to be real) and a blackhole firewall rule that causes ssh connection attempts to time out.

as far as I can tell, no nodes are required, and anything that makes the service connection attempt fail fast would suffice (e.g. reference a service with no endpoints)

**Comment by cheftako** [comment]
On Mon, Oct 29, 2018 at 2:14 PM Jordan Liggitt <notifications@github.com>
wrote:

> Also, do we have way of creating a test that simulates this condition?
> I.e. many nodes (they don't have to be real) and a blackhole firewall rule
> that causes ssh connection attempts to time out.
>
> as far as I can tell, no nodes are required, and anything that makes the
> service connection attempt fail fast would suffice (e.g. reference a
> service with no endpoints)
>
When I was working on the https tunnel patch I on several occasions pointed
the KAS at invalid IP addresses (or valid IP addresses with bad certs) and
did not see the a CPU usage problem.

**Comment by yue9944882** [comment]
OMG, i hit this issue too! 😂

the log shows that the available controller is retrying frequently on error (of which the cause we could ignore in my case).

```
E1130 06:48:09.711989       1 available_controller.go:317] XOVKOVCN/default/default/v1beta1.authorization.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.authorization.k8s.io": StorageError: invalid object, Code: 4, Key: /registry/apiregistration.k8s.io/apiservices/v1beta1.authorization.k8s.io, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 346494b6-d831-11e8-9e13-00163e0e5c97, UID in object meta: 2d69a274-f3b7-11e8-a70a-00163e0f12ad
E1130 06:48:34.349474       1 available_controller.go:317] XOVKOVCN/default/default/v1beta1.rbac.authorization.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.rbac.authorization.k8s.io": StorageError: invalid object, Code: 4, Key: /registry/apiregistration.k8s.io/apiservices/v1beta1.rbac.authorization.k8s.io, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 07c0b31b-d82e-11e8-b9aa-00163e1291ae, UID in object meta: 3618a146-f3b7-11e8-9636-00163e0e5c97
E1130 06:48:34.398878       1 available_controller.go:317] XOVKOVCN/default/default/v1beta1.authorization.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.authorization.k8s.io": StorageError: invalid object, Code: 4, Key: /registry/apiregistration.k8s.io/apiservices/v1beta1.authorization.k8s.io, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 346494b6-d831-11e8-9e13-00163e0e5c97, UID in object meta: 2d69a274-f3b7-11e8-a70a-00163e0f12ad
E1130 06:48:39.709118       1 available_controller.go:317] XOVKOVCN/default/default/v1beta1.rbac.authorization.k8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.rbac.authorization.k8s.io": StorageError: invalid object, Code: 4, Key: /registry/apiregistration.k8s.io/apiservices/v1beta1.rbac.authorization.k8s.io, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 07c0b31b-d82e-11e8-b9aa-00163e1291ae, UID in object meta: 3618a146-f3b7-11e8-9636-00163e0e5c97
E1130 06:48:39.721236       1 available_controller.go:317] XOVKOVCN/default/default/v1beta1.authorization.ka8s.io failed with: Operation cannot be fulfilled on apiservices.apiregistration.k8s.io "v1beta1.authorization.k8s.io": StorageError: invalid object, Code: 4, Key: /registry/apiregistration.k8s.io/apiservices/v1beta1.authorization.k8s.io, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: 346494b6-d831-11e8-9e13-00163e0e5c97, UID in object meta: 2d69a274-f3b7-11e8-a70a-00163e0f12ad
```
